File: built-in plugins dsp/mb-mono-maker.cpp
/**
 * MB Mono Maker
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Collapse low frequencies to mono for tighter bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MONO_MAKER_H
#define MB_MONO_MAKER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMonoMaker : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mono-maker";
    static constexpr const char* PLUGIN_NAME    = "MB Mono Maker";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float frequency = 120f;  // range [20, 500]
    float slope = 12f;  // range [6, 48]
    float mix = 1f;  // range [0, 1]
    };

    MbMonoMaker() = default;
    ~MbMonoMaker() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.frequency = std::clamp(params.frequency, 20f, 500f);
        params.slope = std::clamp(params.slope, 6f, 48f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Mono Maker
        return input;
    }
};

#endif // MB_MONO_MAKER_H

File: built-in plugins dsp/mb-mono-maker.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMonoMakerPlugin: PluginDefinition = { id: 'mb-mono-maker', slug: 'mb-mono-maker', name: 'MB Mono Maker', category: 'effect', type: 'stereo' as any, version: '1.0.0', description: 'Collapse low frequencies to mono for tighter bass', author: 'Max Booster', grade: 'A', parameters: [{ id: 'frequency', name: 'Frequency', type: 'float', defaultValue: 120, minValue: 20, maxValue: 500, automatable: true }, { id: 'slope', name: 'Slope', type: 'float', defaultValue: 12, minValue: 6, maxValue: 48, automatable: false }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { frequency: 120, slope: 12, mix: 1 } };

export default MbMonoMakerPlugin;
